lib/blockchain/chain.js

```diff
@@ -950,6 +956,46 @@ class Chain extends AsyncEmitter {
     return this.emitAsync('reorganize', tip, competitor);
   }
 
+  /**
+   * Checks if a reorganization breaks the set nlrLimit,
+   * if it does, it invalidates necessary blocks and returns true
+   * else return false
+   * @param {ChainEntry} tip - Current tip of this chain.
+   * @param {ChainEntry} fork - The tip of the fork.
+   * @param {ChainEntry} competitor - The competing chain's tip.
+   * @returns {Promise<boolean>}
+   */
+  async noLongReorg (tip, fork, competitor) {
+    if (tip.height - fork.height >= this.network.nlrLimit) {
+      if (this.network.nlrLimit !== 0) {
+        this.logger.warning(
+          'NLR Activated: current=%h(%d) fork=%h(%d) reorg_size=%d',
+          tip.hash,
+          tip.height,
+          competitor.hash,
+          competitor.height,
+          tip.height - fork.height
+        );
+
+        // mark invalid_child from tip of fork to second block of fork
+        while (competitor.height > fork.height + 2) {
+          const previous = await this.getPrevious(competitor)
+          this.invalidate(previous.hash)
```

Comment:
I think `invalidate` needs an await since it's also an async function pending a lock on the chain